### ISSUE TITLE
Historical information on Autoconfig

### ISSUE THREAD

**Issue by ben221199** [benbucksch/autoconfig-spec#8]
# Versions

There are 2 versions of Autoconfig: `1.0` and `1.1`.

Autoconfig 1.0 was supported in ThunderBird 3.0.
Autoconfig 1.1 is supported in ThunderBird 3.1 and later.

# URLs

Each version has its own URLs to request XML data.

For `1.0` this is:
- `https://autoconfig.$DOMAIN/mail/mozilla.xml?emailaddress=$EMAILADDRESS`

For `1.1` this is:
- `http://autoconfig.$DOMAIN/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS`
- `http://$DOMAIN/.well-known/autoconfig/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS` (As fallback)

NOTE: The variable `emailaddress` is sent, containing the e-mailaddress.

# ISPDB URLs

There is also a central ISP database hosted by ThunderBird that contains some domains. This URL is saved in the `mailnews.auto_config_url` preference.

For `1.0` this is:
- `https://live.mozillamessaging.com/autoconfig/`

For `1.1` this was:
- `https://live.mozillamessaging.com/autoconfig/v1.1/`

But now it is:
- `https://live.thunderbird.net/autoconfig/v1.1/`

NOTE: The variable `emailaddress` will not be send to the ISPDB.

NOTE: Since some version 3.0.x (I have to investigate further), it is possible to use `{{domain}}`. If present, ThunderBird will replace this variable with the domain name. Else, it will append it. There is also a variable `{{accounts}}`. That one will always be replaced if present.

# XML

Both versions use XML for sending their information. There are some difference that make them incompatible with each other.

- `<clientConfig>` - The root element. Both versions present.
  - `<emailProvider>` - The email provider element. Inside `<clientConfig>`. Only once. Both versions present.
    - `<id>` - The email provider ID. In 1.0 it is an element inside `<emailProvider>`. Only once. In 1.1, the element is gone. It is now an attribute on `<emailProvider>`.
    - `<displayName>` - The display name of the email provider. Inside `<emailProvider>`. Only once. Both versions present.
    - `<domain>` - The domain element. Inside `<emailProvider>`. Multiple possible. Both versions present.
    - `<incomingServer>` - An incoming server. Has a `type` attribute. Only `pop3`, `nntp` or `imap` allowed. Inside `<emailProvider>`. Multiple possible in 1.1, not in 1.0. Both versions present.
      - `<hostname>` - Hostname of the server. Inside `<incomingServer>`. Only once. Both versions present.
      - `<port>` - Port of the server. Inside `<incomingServer>`. Only once. Both versions present.
      - `<username>` - Username to use to connect to the server. Inside `<incomingServer>`. Only once. Both versions present.
      - `<socketType>` - Socket type. A value of `plain`, `SSL` or `STARTTLS`. Inside `<incomingServer>`. Only once in 1.0. Multiple possible in 1.1. Both versions present.
      - `<authentication>` - Authentication. For 1.0, a value of `plain` or `secure`. For 1.1, a value of `password-cleartext`, `plain`, `password-encrypted`, `secure`, `GSSAPI` or `NTLM`, where `plain` and `secure` are deprecated. Inside `<incomingServer>`. Only once in 1.0. Multiple possible in 1.1. Both versions present.
      - `<pop3>` - Additional POP3 options. Inside `<incomingServer>`. Only once. Both versions present. Only useful when type is POP3.
        - `<leaveMessagesOnServer>` - Leave messages on server. Boolean. Inside `<pop3>`. Only once. Both versions present.
        - `<daysToLeaveMessagesOnServer>` - Days to leave messages on the server. Integer. Inside `<pop3>`. Only once. Both versions present.
        - `<downloadOnBiff>` - (I don't know what this does, but something with downloading) Boolean. Inside `<pop3>`. Only once. Both versions present.
    - `<outgoingServer>` - An outgoing server. Has a `type` attribute. Only `smtp` allowed. Inside `<emailProvider>`. Multiple possible in 1.1, not in 1.0. Both versions present.
      - `<hostname>` - Hostname of the server. Inside `<outgoingServer>`. Only once. Both versions present.
      - `<port>` - Port of the server. Inside `<outgoingServer>`. Only once. Both versions present.
      - `<username>` - Username to use to connect to the server. Inside `<outgoingServer>`. Only once. Both versions present.
      - `<socketType>` - Socket type. A value of `plain`, `SSL` or `STARTTLS`. Inside `<outgoingServer>`. Only once in 1.0. Multiple possible in 1.1. Both versions present.
      - `<authentication>` - Authentication. For 1.0, a value of `none`, `plain`, `secure` or `smtp-after-pop`. For 1.1, a value of `none`, `client-IP-address`, `smtp-after-pop`, `password-cleartext`, `plain`, `password-encrypted`, `secure`, `GSSAPI` or `NTLM`, where `plain` and `secure` are deprecated. Inside `<outgoingServer>`. Only once in 1.0. Multiple possible in 1.1. Both versions present.
      - `<addThisServer>` - Field to add this server to the SMTP server list. Contains a boolean. Inside `<outgoingServer>`. Only once. Both versions present.
      - `<useGlobalPreferredServer>` - Field to use global preferred server of the SMTP server list. Contains a boolean. Inside `<outgoingServer>`. Only once. Both versions present.
    - `<inputField>` - An input field. Has a `key` attribute and a `label` attribute. The value is inside the element. Inside `<emailProvider>`. Multiple possible. Both versions present. Optional.

NOTE: In some later versions I came across the `exchange` type for incoming servers (including `<owaURL>`, `<ewsURL>` and `<easURL>`), the `<useGlobalPreferredServer>` inside an incoming server, the `<password>` element in both incoming and outgoing servers and the `OAuth2` authentication option. For now I see them as non-1.1 and I don't list them in the list above.

NOTE: It doesn't seem that ThunderBird checks version. For now, I didn't list the `version` attribute on `<clientConfig>`.

NOTE: I didn't yet specify which elements are required and which are optional.

NOTE: I didn't yet specify which elements support variables and will possibily have some parts replaced by real values.

**Comment by ben221199** [comment]
With the information above and after having done some analysis of the code, I think the RFC should define a new version: `1.2`. Why? Because ThunderBird checks and throws on some values at the moment and in the past, so extending 1.0 and/or 1.1 isn't a simple option.

Apart from specifying 1.2, I think the RFC should also give some attention to 1.0 and 1.1. Because ThunderBird doesn't check `version` and `xmlns`, I think we could make really nice section for it. Change some behaviour, but make it compatible with older clients. Some ideas:

# URLs

At the moment, `1.1` uses the following 2 URLs:
- `http://autoconfig.$DOMAIN/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS`
- `http://$DOMAIN//.well-known/autoconfig/mail/config-v1.1.xml?emailaddress=$EMAILADDRESS`

For `1.0`, we do to the same:
- `http://autoconfig.$DOMAIN/mail/config-v1.0.xml?emailaddress=$EMAILADDRESS`
- `http://$DOMAIN//.well-known/autoconfig/mail/config-v1.0.xml?emailaddress=$EMAILADDRESS`

What about `https://autoconfig.$DOMAIN/mail/mozilla.xml?emailaddress=$EMAILADDRESS`? For backward compatibility with old ThunderBird, we make it redirect to one of the NEW 1.0 URLs. (Or we server the same content. The server owner is the boss.) New clients should use the new URLs.

# XML

For both versions, DTDs and XSDs will be written. Invalid client configs will fail validation. Valid client configs will pass. Because ThunderBird doesn't check `version` and `xmlns` attributes, we can define them with retroactivity. In 1.0, the `version` attrribute is optional and if present MUST contain `1.0`. In 1.1 the attribute must always be present and be `1.1`. For `xmlns` we will invent an URN XMLNS and register it at IANA. For example: `urn:ietf:params:xml:ns:autoconfig-1.0` and `urn:ietf:params:xml:ns:autoconfig-1.1`. Values and elements added over time will possibly not be allowed.

# Version 1.2

What about 1.2? We should design it with extendibility in mind. Of course we can define a version 1.3, but I think that is not best practice if a client has to check all versions backwards to find the client config served by a server that it supports. With 1.2 defined, it client has to check 6 URLs (possibly 7 if it also checks `mozilla.xml`). Lets keep it at that amount by doing some smart things.

So, the URLs will be:
- `http://autoconfig.$DOMAIN/mail/config-v1.2.xml?emailaddress=$EMAILADDRESS`
- `http://$DOMAIN//.well-known/autoconfig/mail/config-v1.2.xml?emailaddress=$EMAILADDRESS`

The version attribute on `<clientConfig>` will be:
- `<clientConfig version="1.2">`

And the XMLNS will be (as example):
- `urn:ietf:params:xml:ns:autoconfig-1.2`

**Comment by ben221199** [comment]
I saw the latest draft had `1.2` in its version due to https://github.com/benbucksch/autoconfig-spec/commit/4fa94ced85ba2270ffefb18b286fe5db10b6bd05. That is cool.

**Comment by ben221199** [comment]
The minimal config file for 1.0 is this:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<clientConfig><!--Element name is not checked-->
	<emailProvider><!--Throwing doesn't work when element is absent, but an exception will occur due to missing children-->
		<incomingServer type="imap"><!--Only checks for "type" attribute having one of the three values.-->
			<hostname>imap.googlemail.com</hostname><!--Required to be filled during variable replacement-->
			<socketType>SSL</socketType><!--Checks if having one of the three values.-->
			<authentication>plain</authentication><!--Checks if having one of the two values.-->
		</incomingServer>

		<outgoingServer type="smtp"><!--Only checks for "type" attribute having "smtp".-->
			<hostname>smtp.googlemail.com</hostname><!--Required to be filled during variable replacement-->
			<socketType>SSL</socketType><!--Checks if having one of the three values.-->
			<authentication>plain</authentication><!--Checks if having one of the four values.-->
		</outgoingServer>
	</emailProvider>
</clientConfig2>

```

**Comment by benbucksch** [comment]
Hey,

thanks for the archeological info on autoconfig v1.0. The main difference was support for oAuth2 auth method, IIRC, which required us to allow a fallback auth method, which required us to make auth a 1:n relation (multiple entries), whereas v1.0 allowed only 1 auth method per server.

However, it is irrelevant today, because only Thunderbird 3.0 supported v1.0, and that is now very dead (since over 15 years), so it's not worth documenting or discussing it.

In the process of standardizing, we'll likely make a few backwards-compatible changes and call it v1.2, but still on the old URLs, if they are widely used.

Ben Bucksch

**Comment by benbucksch** [comment]
P.S. Are you Ben Campbell from Thunderbird? You are appearing only as "Ben", which is confusing, because that's also me. Would you mind adding some other word (not number) to your name, so we can differentiate the various persons?

**Comment by ben221199** [comment]
> P.S. Are you Ben Campbell from Thunderbird? You are appearing only as "Ben", which is confusing, because that's also me. Would you mind adding some other word (not number) to your name, so we can differentiate the various persons?

I'm not. I'm a Ben from The Netherlands. I never worked for Thunderbird. I can look into adding my lastname.

**Comment by benbucksch** [comment]
It's an important goal of this standard to stay compatible with existing configs at ISPs. We control the ISPDB, but not the ISPs. While being logical and clean is very important to me, being compatible is more important.

I'll need to do a survey to see which URLs matter in practice. The choice will depend on that.

v1.0 is really dead since a very long time, and existed only for a few months, but even worth an appendix. v1.1 was essentially a bugfix. I do not want people to feel that they should implement v1.0.

Did you implement autoconfig in some software?

Thanks for the XSDs and other validators. That's useful.

**Comment by benbucksch** [comment]
sorry, typo: ... *not* even worth....

I wrote:
> v1.0 is really dead since a very long time, and existed only for a few months, but even worth an appendix.

**Comment by benbucksch** [comment]
Hey @ben221199 , if you have an XSD for 1.2, which the specification specifies, I'd be very happy. Please feel free to open another PR for that.

I think we can close this now. The info will be archived as part of this issue.

**Comment by ben221199** [comment]
My reaction on #9:
> Hi @benbucksch, sorry for my inactivity last months. I had to focus on some other projects. I will finish 1.1 and add 1.2 as soon as I have time.

Also, if possible, I would prefer reopening #9 when I'm done commiting, because it is the original PR with all conversations.
